app/services/foreman_discovery/host_converter.rb

```diff
@@ -21,10 +21,10 @@ def self.to_managed(original_host, set_managed = true, set_build = true, added_a
   def self.set_build_clean_facts(host)
     # fact cleaning
     if Setting['discovery_clean_facts']
-      # clean all facts except those starting with "discovery_"
+      # clean all facts except those from Discovery
       host.define_singleton_method(:clear_facts) do
-          keep_ids = FactValue.where("host_id = #{host.id}").joins(:fact_name).where("fact_names.name like 'discovery_%'").pluck("fact_values.id")
-          FactValue.where("host_id = #{host.id} and id not in (?)", keep_ids).delete_all
+        keep_ids = FactValue.where(host_id: host.id, fact_names: { type: 'DiscoveryFactName' }).joins(:fact_name)
+        FactValue.where.not(id: keep_ids).delete_all
       end
     else
       # clean no facts (default behavior)
```

Comment:
Apologies I misled you here, actually purpose of discovery_clean_facts is to clean facts starting with discovery_ which are some facts which can be useful during provisioning. Fact type is actually irrelevant, can you put this hunk back as it was? Of course the code must search DiscoveryFactName but also starting with discovery_.

Sorry my bad, I told you to do this change but now I see it is incorrect.

Reply:
Nono, what I mean is: delete all facts which are type of DiscoveryFactName AND they start with "discovery_" prefix.